lib/Frontend/Frontend.cpp

```diff
@@ -509,17 +506,34 @@ CompilerInstance::getInputBufferAndModuleDocBufferIfPresent(
   if (!inputFileOrErr) {
     Diagnostics.diagnose(SourceLoc(), diag::error_open_input_file, input.file(),
                          inputFileOrErr.getError().message());
-    return std::make_pair(nullptr, nullptr);
+    return None;
   }
   if (!serialization::isSerializedAST((*inputFileOrErr)->getBuffer()))
-    return std::make_pair(std::move(*inputFileOrErr), nullptr);
+    return ModuleBuffers(std::move(*inputFileOrErr));
 
-  if (Optional<std::unique_ptr<llvm::MemoryBuffer>> moduleDocBuffer =
-          openModuleDoc(input)) {
-    return std::make_pair(std::move(*inputFileOrErr),
-                          std::move(*moduleDocBuffer));
-  }
-  return std::make_pair(nullptr, nullptr);
+  return ModuleBuffers(
+    std::move(*inputFileOrErr),
+    openModuleDoc(input).getValueOr(nullptr),
+    openModuleSourceInfo(input).getValueOr(nullptr)
+  );
+}
+
+Optional<std::unique_ptr<llvm::MemoryBuffer>>
+CompilerInstance::openModuleSourceInfo(const InputFile &input) {
+  llvm::SmallString<128> moduleSourceInfoFilePath(input.file());
+  llvm::sys::path::replace_extension(moduleSourceInfoFilePath,
+                                     file_types::getExtension(file_types::TY_SwiftSourceInfoFile));
+  std::string NonPrivatePath = moduleSourceInfoFilePath.str().str();
```

Comment:
Nitpicks: this should be a SmallString too, and it should have a lowerCamelCase name, and I don't love that the name is described in terms of the *other* path.

Reply:
Actually, I don't think we need to do the "Project" thing for the partial modules.